feat(inputs): add text input styles

StommePoes opened the pull request:
I couldn't figure out how to style placeholders without vendor prefixes in a way that works, so they are not styled.

I had to make other changes: @umahaea @aaronkaka please check.

parkerpearson commented:
Do you mean the placeholders are not italicized? Do we also lose control over their color? If so I imagine this is something we'll have to review more with accessibility.

StommePoes commented:
Yeah we lose that because either a) I can't have vendor prefixes due to our linter (and *generally* that's a good thing, just sucks here) and b) the solutions I've found via SO and the sass site are adding a space between the words "input" and "::-webkit..." etc. I believe the space is why our styles aren't being applied.

Also I had talked earlier with some other (non-Pearson) a11y tweeps about styling placeholders and people have been doing similar (#737373 colour and italicising them), mostly with the idea that for those people who *can* see text at that contrast, the italicising can help against the visual same-ness with input values, thus helping users not get tricked into thinking the placeholders are default values who may be submitted.
Granted that's only if the user can see and notice that difference.

There's also some other stuff Parker, I added a class for a wrapping div where a developer can decide if label-input pairs may be inline or stacked, and added a (temporary) bottom margin of 1em for the div around label-input paris so there's some room between them. Currently multiple inputs together isn't yet in the design page so I figure we can change those as they appear.
But wanted some eyes on this before merging because I had done that.

parkerpearson commented:
So is the only reason we can't style the placeholder is because our linter
doesn't recognize the rules? Or do the vendor prefixes not cover our set of
supported browsers?

For the inline style, do you mean the label is still stacked above the
input, it's just the sequence of input/label pairs that are inline?

I think that's fine on the temporary spacing, we'll finalize that in a
forms component.

On Sat, Apr 16, 2016 at 2:04 AM, Mallory <notifications@github.com> wrote:

> Yeah we lose that because either a) I can't have vendor prefixes due to
> our linter (and *generally* that's a good thing, just sucks here) and b)
> the solutions I've found via SO and the sass site are adding a space
> between the words "input" and "::-webkit..." etc. I believe the space is
> why our styles aren't being applied.
>
> Also I had talked earlier with some other (non-Pearson) a11y tweeps about
> styling placeholders and people have been doing similar (#737373 colour and
> italicising them), mostly with the idea that for those people who *can*
> see text at that contrast, the italicising can help against the visual
> same-ness with input values, thus helping users not get tricked into
> thinking the placeholders are default values who may be submitted.
> Granted that's only if the user can see and notice that difference.
>
> There's also some other stuff Parker, I added a class for a wrapping div
> where a developer can decide if label-input pairs may be inline or stacked,
> and added a (temporary) bottom margin of 1em for the div around label-input
> paris so there's some room between them. Currently multiple inputs together
> isn't yet in the design page so I figure we can change those as they appear.
> But wanted some eyes on this before merging because I had done that.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/Pearson-Higher-Ed/elements/pull/61#issuecomment-210760999>
>

StommePoes commented:
>"So is the only reason we can't style the placeholder is because our linter
doesn't recognize the rules? Or do the vendor prefixes not cover our set of
supported browsers?"
The linter disallows vendor prefixes. The usual way around this apparently is to use mixins, which are exempt from that linting rule-- however I haven't figured out how to make this mixin add to a rule (instead of just adding a cluster of CSS properties like most of our mixins do). So I've got the mixin commented out because when it runs, instead of producing this CSS
`input:disabled::-webkit-placeholder-blah-blah {`
it produces this
`input:disabled: ::-webkit-placeholder-blah-blah {`
with that space. So currently with this code, we're seeing the browser defaults because with the code with the space, I can't override their internal stylesheets.

>"For the inline style, do you mean the label is still stacked above the
input, it's just the sequence of input/label pairs that are inline?"
Yes: Gayan originally had the two basic inputs next to each other (normal and small) but they were without labels. In order to get the labels above but the label-input pairs together, we need to wrap them in something. Since I had Sireesha by default wrap her label-input pairs for checkboxes and radios in divs, I figured for now (so, temporary until you/Ed decide what you actually want) there's a class that makes those divs inline/inline-block and stack.

So basically I was keeping Gayan's setup which I think he took from your design sheet where you have a side-by-side to show better the difference between normal and small inputs.

I consider this temporary as well since we don't actually have designs for multiple label-input pairs (a full form of any sort). Same as the temporary space under stacked label-input pairs-- the only thing I would require a11y-wise for either is that when they're stacked, the label is visually close enough to its input that it's clear who it belongs to (so greater space between pairs than between a label and its input) and same for side-by-side when they are checks/radios-- that it's visually clear which label belongs to which check/radio. For now, I used 1em but you guys can decide if you want greater spaces between.
Another thing to consider if a designer/developer needs columns, if elements may or may not have help/error text underneath, the two sides may not line up unless that space is default... and I dunno if we want default minimum 2 lines of help/error text room under each and every label-input pair.

All way way in the future...
